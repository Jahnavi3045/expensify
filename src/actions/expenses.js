import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';
import { push, ref ,onValue,get, remove, update} from 'firebase/database';
import { useDispatch } from 'react-redux';

export const addExpense=(expense)=>({
    type:'ADD_EXPENSE',
    expense   
});

export const startAddExpense=(expenseData={})=>{
    // const dispatch=useDispatch();
    return(dispatch)=>{
        const {
            description='',
            note='',
            amount=0,
            createdAt=0
        }=expenseData;

        const expense={description,note,amount,createdAt}
        return push(ref(database,'expense'),expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    }
}
export const removeExpense=({id}={})=>({
    type:'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense=({id}={})=>{
    return (dispatch)=>{
        return remove(ref(database,`expense/${id}`)).then(()=>{
            dispatch(removeExpense({id}))
        })
    }
}

export const editExpense=(id,updates)=>({
    type:'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense=(id,updates)=>{
    return(dispatch)=>{
        return update(ref(database,`expense/${id}`),updates).then(()=>{
            dispatch(editExpense(id,updates));
        })
    }
}

export const setExpense=(expenses)=>({
    type:'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return async (dispatch) => {
      try {
        const expenses = [];
        const snapshot = await get(ref(database, 'expense')); // Use get() to fetch once
        
        if (snapshot.exists()) {
            snapshot.forEach((childSnapshot) => {
              expenses.push({
                id: childSnapshot.key,
                ...childSnapshot.val(),
              });
            });
            console.log(expenses)
            dispatch(setExpense(expenses));
            console.log(expenses)
        } 
      } catch (error) {
        console.error('Error fetching expenses:', error);
      }
    };
  };