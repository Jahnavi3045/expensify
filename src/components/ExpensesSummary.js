// import React from 'react'
// import getVisibleExpenses from '../selectors/expenses'
// import expenses_total from '../selectors/expenses-total'
// import numeral from 'numeral';
// import { useSelector } from 'react-redux'

// const ExpensesSummary = () => {
//     const visibleExpenses=useSelector((state)=>getVisibleExpenses(state.expenses,state.filters))
//     const expensesTotal=useSelector((state)=>expenses_total(visibleExpenses))
//     const expenseCount=visibleExpenses.length
//     const expenseWord=expenseCount===1?'expense':'expenses'
//     const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

//   return (
//     <div>
//       viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
//     </div>
//   )
// }

// export default ExpensesSummary

import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import getVisibleExpenses from '../selectors/expenses'
import expenses_total from '../selectors/expenses-total'

const ExpensesSummary = ({expenseCount,expensesTotal}) => {
    const expenseWord=expenseCount===1?'expense':'expenses'
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
      viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
    </div>
  )
}

const mapStateToProps=(state)=>{
    const visibleExpenses=getVisibleExpenses(state.expenses,state.filters)

    return{
        expenseCount:visibleExpenses.length,
        expensesTotal:expenses_total(visibleExpenses)
    }
}
export default connect(mapStateToProps)(ExpensesSummary)

