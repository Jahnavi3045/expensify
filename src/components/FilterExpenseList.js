// import React from 'react'
// import { DateRangePicker } from 'react-dates';
// import {useSelector,useDispatch} from 'react-redux'
// import { setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../actions/filters'


// const FilterExpenseList = () => {
//   state={
//     calendarFocused:false
//   }
//     const text=useSelector((state)=>state.filters.text);
//     const sortBy=useSelector((state)=>state.filters.sortBy);
//     const startDate=useSelector((state)=>state.filters.startDate);
//     const endDate=useSelector((state)=>state.filters.endDate);
//     const focusedInput=useSelector((state)=>state.calendarFocused)
//     const dispatch=useDispatch();

//     const handleTextChange = (e) => {
//         dispatch(setTextFilter(e.target.value)); 
//     };
//     const sortByChange=(e)=>{
//       if(e.target.value==='date'){
//         dispatch(sortByDate())
//       }
//       else if(e.target.value==='amount'){
//         dispatch(sortByAmount())
//       }
//     }
//     const onDatesChange=({startDate,endDate})=>{
//       dispatch(setStartDate(startDate))
//       dispatch(setEndDate(endDate))
//     }
//     const onFocusChange=(calendarFocused)=>{
//       this.setState(()=>({calendarFocused}))
//     }
//     return (
//       <div>
//         <input
//           type="text"
//           value={text}
//           onChange={handleTextChange}
//           placeholder="Filter expenses"
//         />
//         <select value={sortBy} onChange={sortByChange}>
//           <option value={'date'}>Date</option>
//           <option value={'amount'}>Amount</option>
//         </select>
//         <DateRangePicker
//           startDate={startDate}
//           endDate={endDate}
//           onDatesChange={onDatesChange}
//           focusedInput={focusedInput}
//           onFocusChange={onFocusChange}
//         />
//       </div>
//     );
//   };
  

// export default FilterExpenseList


// import React, { Component } from 'react';
// import { DateRangePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css'; // Import styles
// import 'react-dates/initialize';
// import { connect } from 'react-redux';
// import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

// class FilterExpenseList extends Component {
//   // constructor(props) {
//   //   super(props);
//   //   this.state = {
//   //     calendarFocused: null,
//   //     startDate:moment().startOf('month'), // Class component state for calendar focus
//   //     endDate:moment().endOf('month') // Class component state for calendar focus
//   //   };
//   // }
//   state={
//     calendarFocused:null
//   }

//   handleTextChange = (e) => {
//     this.props.setTextFilter(e.target.value);
//   };

//   sortByChange = (e) => {
//     if (e.target.value === 'date') {
//       this.props.sortByDate();
//     } else if (e.target.value === 'amount') {
//       this.props.sortByAmount();
//     }
//   };

//   onDatesChange = ({ startDate, endDate }) => {
//     this.props.setStartDate(startDate);
//     this.props.setEndDate(endDate);
//     // this.setState(()=>({startDate,endDate}))
//   };

//   onFocusChange = (calendarFocused) => {
//     this.setState(() => ({ calendarFocused }));
//   };

//   render() {
//     return (
//       <div>
//         <input
//           type="text"
//           value={this.props.text}
//           onChange={this.handleTextChange}
//           placeholder="Filter expenses"
//         />
//         <select value={this.props.sortBy} onChange={this.sortByChange}>
//           <option value="date">Date</option>
//           <option value="amount">Amount</option>
//         </select>
//         <DateRangePicker
//           startDate={this.props.filters.startDate}
//           endDate={this.props.filters.endDate}
//           onDatesChange={this.onDatesChange}
//           focusedInput={this.state.calendarFocused}
//           showClearDates={true}
//           onFocusChange={this.onFocusChange}numberOfMonths={1}
//           isOutsideRange={()=>false}
//         />
//       </div>
//     );
//   }
// }

// // Map Redux state to props
// const mapStateToProps = (state) => ({
//   text: state.filters.text,
//   sortBy: state.filters.sortBy,
//   startDate: state.filters.startDate,
//   endDate: state.filters.endDate,
// });

// // Map Redux actions to props
// const mapDispatchToProps = (dispatch) => ({
//   setTextFilter: (text) => dispatch(setTextFilter(text)),
//   sortByDate: () => dispatch(sortByDate()),
//   sortByAmount: () => dispatch(sortByAmount()),
//   setStartDate: (startDate) => dispatch(setStartDate(startDate)),
//   setEndDate: (endDate) => dispatch(setEndDate(endDate)),
// });

// // Connect the class component to Redux
// export default connect(mapStateToProps, mapDispatchToProps)(FilterExpenseList);



import React, { useState } from 'react';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css'; // Import styles
import 'react-dates/initialize';
import { useDispatch, useSelector } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

const FilterExpenseList = () => {
  const [calendarFocused, setCalendarFocused] = useState(null); // Replacing class state
  const dispatch = useDispatch(); // Use dispatch hook to dispatch actions

  // Extract filters state from Redux using useSelector
  const filters = useSelector((state) => state.filters);

  const handleTextChange = (e) => {
    dispatch(setTextFilter(e.target.value));
  };

  const sortByChange = (e) => {
    if (e.target.value === 'date') {
      dispatch(sortByDate());
    } else if (e.target.value === 'amount') {
      dispatch(sortByAmount());
    }
  };

  const onDatesChange = ({ startDate, endDate }) => {
    dispatch(setStartDate(startDate));
    dispatch(setEndDate(endDate));
  };

  const onFocusChange = (calendarFocused) => {
    setCalendarFocused(calendarFocused); // Update the calendar focus state
  };

  return (
    <div>
      <input
        type="text"
        value={filters.text}
        onChange={handleTextChange}
        placeholder="Filter expenses"
      />
      <select value={filters.sortBy} onChange={sortByChange}>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker
        startDate={filters.startDate}
        endDate={filters.endDate}
        onDatesChange={onDatesChange}
        focusedInput={calendarFocused}
        showClearDates={true}
        onFocusChange={onFocusChange}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
    </div>
  );
};

export default FilterExpenseList;

