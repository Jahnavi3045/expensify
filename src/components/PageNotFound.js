import React from 'react'
import { Link } from 'react-router-dom'

const PageNotFound = () => {
  return (
    <div>
      404-PAGE NOT FOUND
      <Link to='/'>Go to Home page</Link>
    </div>
  )
}

export default PageNotFound
