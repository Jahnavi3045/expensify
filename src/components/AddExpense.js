import React from 'react'
import { useNavigate } from 'react-router-dom'
import ExpenseForm from './ExpenseForm'
import { useDispatch } from 'react-redux'
import {startAddExpense} from '../actions/expenses'

const AddExpense = (props) => {
  console.log(props)
  const dispatch=useDispatch()
  const navigate=useNavigate()
  return (
    <div>
      Add expense
      <ExpenseForm
        onSubmit={(expense)=>{
          dispatch(startAddExpense(expense))
          navigate('/')
        }}
      />
    </div>
  )
}

export default AddExpense
