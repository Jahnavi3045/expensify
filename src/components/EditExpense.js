import React from 'react'
import { useNavigate } from 'react-router-dom'
import { useParams } from 'react-router'
import ExpenseForm from './ExpenseForm'
import { useDispatch, useSelector } from 'react-redux'
import { startEditExpense,startRemoveExpense } from '../actions/expenses'

const EditExpense = () => {
    // console.log(props)
    const {id}=useParams()
    const expense=useSelector((state)=>state.expenses.find((expense)=>expense.id===id))
    const dispatch=useDispatch()
    const navigate=useNavigate()
    console.log(expense)

    const selectExpense=(expense)=>{
      console.log(id)
      dispatch(startEditExpense(id,expense))
      console.log(expense)
      navigate('/')
    }

  return (
    <div>
      <ExpenseForm
        expense={expense}
        onSubmit={selectExpense}
      />
      <button onClick={()=>{
        dispatch(startRemoveExpense ({id}))
        navigate('/')
      }}>Remove</button>
    </div>
  )
}

export default EditExpense
