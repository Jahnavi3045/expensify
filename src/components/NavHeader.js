import React from 'react'
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux'
import { startLogOut } from '../actions/auth'


export const NavHeader = ({startLogOut}) => {
  return (
    <div>
        <h2>Expensify</h2>
        <NavLink to={'/'}>Home</NavLink>
        <NavLink to={'/addExpense'}>addExpense</NavLink>
        <NavLink to={'/help'}>Help</NavLink>
        <button onClick={startLogOut}>log out</button>
    </div>
  )
}

const mapDispatchToProps=(dispatch)=>({
  startLogOut:()=>dispatch(startLogOut())
})
export default connect(undefined,mapDispatchToProps)(NavHeader);
