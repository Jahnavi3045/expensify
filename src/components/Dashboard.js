import React from 'react'
// import { useDispatch, useSelector } from 'react-redux'
// import { useDispatch} from 'react-redux'
// import {addExpense} from '../actions/expenses'
// import {setTextFilter} from '../actions/filters'
// import getVisibleExpenses from '../selectors/expenses'
import ExpenseList from './ExpenseList'
import FilterExpenseList from './FilterExpenseList'
import ExpensesSummary from './ExpensesSummary'

const Dashboard = () => {
  // const dispatch=useDispatch();
  // dispatch(addExpense({description:'water bill',amount:355,createdAt:110}));
  // dispatch(addExpense({description:'gas bill',createdAt:100,amount:1000}));
  // dispatch(addExpense({description:'wifi bill',createdAt:35,amount:200}));
  // dispatch(addExpense({description:'house rent',createdAt:0,amount:6000}));
  // dispatch(setTextFilter('gas'));
  
  // const state=useSelector((state)=>state)
  // const visibleExpenses=getVisibleExpenses(state.expenses,state.filters)
  // console.log(visibleExpenses)

  return (
    <div>
      <FilterExpenseList/>
      <ExpenseList/>
      <ExpensesSummary/>
    </div>
  )
}

export default Dashboard
