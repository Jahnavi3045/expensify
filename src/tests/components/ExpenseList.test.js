import React from "react";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from '../fixtures/expenses'
import ReactShallowRenderer from 'react-test-renderer/shallow'

test('should render expense list correctly',()=>{
    const renderer=new ReactShallowRenderer()
    renderer.render(<ExpenseList expenses={expenses}/>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
})

test('should render expense list with empty message',()=>{
    const renderer=new ReactShallowRenderer()
    renderer.render(<ExpenseList expenses={[]}/>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
})