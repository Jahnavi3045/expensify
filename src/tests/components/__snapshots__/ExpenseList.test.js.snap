// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list correctly 1`] = `
<div>
  <ExpenseListItem
    amount={1000}
    createdAt={-345600000}
    description="water bill"
    id="1"
  />
  <ExpenseListItem
    amount={800}
    createdAt={0}
    description="wifi bill"
    id="2"
  />
  <ExpenseListItem
    amount={1200}
    createdAt={345600000}
    description="gas bill"
    id="3"
  />
</div>
`;

exports[`should render expense list with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
